README.md
# API Draft #

## Mission updates ##

### Access ###

**Endpoint:** /mission/:learnerId/latest

**Method:** GET

**Notes:** For now learnerId can be any integer. I will generate random content each request.

### Contract ###

```json
{
  "updates": [
    {
        "image": "https://someimageurl/",
        "status": "green",
        "title": "Mission 1 title",
        "link": "https://math-stg.aida.pearson.com/api/v1/assets/videos/467",
    },{
        "image": "https://someimageurl/",
        "status": "yellow",
        "title": "Mission 2 title",
        "link": "https://math-stg.aida.pearson.com/api/v1/assets/videos/467",
    },
  ]
}
```

**Status:** We didn't covered that so I can assume "green", "yellow", "red" for now.

**Images:** Images will be pushed also right away. This content will be using standard HTTP/2 Push.



# Focused Practice #

### Access ###

**Endpoint:** /practice/:learnerId/focused

**Method:** GET

**Notes:** For now learnerId can be any integer. I will generate random content each request.

### Contract ###

```json
{
    "title": "Some Practice title",
    "description": "This is practice description that can be lorem ipsum",
    "icon": "icon_globe",
    "link": "https://math-stg.aida.pearson.com/api/v1/assets/videos/467"
}
```

**Link:** I do believe this widget needs to be clickable and move somewhere - let's link it for now with video that you've provided.


# Content sitemap - Question mark button #

### Access ###

**Endpoint:** /sitemap/

**Method:** GET

**Notes:** This is static content index that can be shown.

### Contract ###

```json
{
  "sitemap": {
    "about_page": {
      "title": "About information",
      "link": "https://testing-stream-service-uqekqerd5a-uc.a.run.app/static/about.html"
    }
  }
}
```

# Notifications #

### Access ###

**Endpoint:** /notification

**Method:** GET

**Notes:** This is notifications HTTP/2 stream endpoint. Notification will be sent after 10 seconds.






CHANGELOG.md
# Changelog #

## v0.2 [2020-05-29] ##

* Changed latest mission updates title from "Talk about gestures" to "Talk about grammar"
* Changed focused practice title from "Talk about gestures" to "Talk about grammar"
* Changed focused practice desctiption from "Talk about getures learns you how to talk about gestures" to more meaningful "Talk about grammar improves your basic grammar skills. You will study basic phonology, morphology and syntax, often complemented by phonetics, semantics and pragmatics".
* Changed notifications API by adding additional message sent after 1 hour from last message sent (if app is still running)
* Added CHANGELOG.md for updates references


main.go
package main

import (
	"github.com/gin-gonic/gin"
	"log"
	"newPocApi/endpoints/mission"
	"newPocApi/endpoints/notifications"
	"newPocApi/endpoints/practice"
	"newPocApi/endpoints/sitemap"
	"newPocApi/middleware"
	"os"
	"path/filepath"
)

var version = "0"

func main() {
	log.Println("Version", version)

	// new gin
	newGin := gin.New()

	newGin.Use(middleware.New(version).Middleware)

	newGin.GET("/info", func(ctx *gin.Context){
		ctx.JSON(200, struct {
			Version 	string	`json:"version"`
		}{
			Version: version,
		})
	})

	mCtrl := mission.NewBinder()
	newGin.GET("/mission/:learnerId/latest", mCtrl.Bind)

	pCtrl := practice.NewBinder()
	newGin.GET("/practice/:learnerId/focused", pCtrl.Bind)

	sCtrl := sitemap.NewBinder()
	newGin.GET("/sitemap", sCtrl.Bind)

	nCtrl := notifications.NewBinder()
	newGin.GET("/notification", nCtrl.Bind)


	newGin.Static("/static/", getLocalPath("static/"))

	newGin.RunTLS(":8443", getLocalPath("domain.crt"), getLocalPath("domain.key"))
}

func getLocalPath(filename string) string {
	abs,_ := filepath.Abs(os.Args[0])
	appPath := filepath.Dir(abs)
	return filepath.Join(appPath, filename)
}


middleware/headershandler.go
package middleware

import (
	"github.com/gin-gonic/gin"
	"strings"
	"time"
)

type HeadersHandler struct {
	version string
}

func (h *HeadersHandler) Middleware(context *gin.Context) {
	headers := map[string]string {
		"Access-Control-Allow-Origin": "*",
		"Access-Control-Allow-Methods": "GET, HEAD",
		"Access-Control-Allow-Headers": "Origin, User-ID, User-Answer, Content-Type",
		"Expire": time.Now().Add(-1*time.Hour*24*365).Format(time.RFC850),
		"Pragma": "no-cache",
		"Cache-Control": "no-cache",
		"X-App-Version": h.version,
	}
	for key,val := range headers {
		context.Writer.Header().Add(key, val)
	}

	if strings.ToUpper(context.Request.Method) == "OPTIONS" {
		context.Status(200)
		context.Done()
	}

	context.Next()
}

func New(version string) *HeadersHandler {
	return &HeadersHandler{
		version: version,
	}
}

endpoints/practice/practice.go
package practice

import (
	"github.com/gin-gonic/gin"
	"newPocApi/endpoints"
)

type practiceDetails struct {
	Title 	string	`json:"title"`
	Description  string	`json:"description"`
	Icon 	string	`json:"icon"`
	Link 	string 	`json:"link"`
}

type ctrl struct {

}

func (m *ctrl) Bind(ctx *gin.Context) {
	ctx.JSON(200, m.getFocused())
	ctx.Done()
}

func (m *ctrl) getFocused() *practiceDetails {
	return &practiceDetails{
		Title:       "Talk about grammar #1",
		Description: "Talk about grammar improves your basic grammar skills. You will study basic phonology, morphology and syntax, often complemented by phonetics, semantics and pragmatics.",
		Icon:        "icon_globe",
		Link:        "https://math-stg.aida.pearson.com/api/v1/assets/videos/467",
	}
}

func NewBinder() endpoints.Binder {
	m := &ctrl{}
	return m
}

endpoints/notifications/notifications.go
package notifications

import (
	"encoding/json"
	"fmt"
	"github.com/gin-gonic/gin"
	"github.com/gorilla/websocket"
	"io"
	"log"
	"newPocApi/endpoints"
	"time"
)

type Update struct {
	Icon 	string	`json:"icon"`
	Description 	string	`json:"description"`
	Title 	string	`json:"title"`
	Link 	string 	`json:"link"`
}

type ctrl struct {

}

func (m *ctrl) BindStream(ctx *gin.Context) {
	ctx.Stream(func(w io.Writer) bool {
		log.Println("Stream opened. Notification will be sent in 10 seconds")
		time.Sleep(10 * time.Second)
		json.NewEncoder(w).Encode(m.getNotification(fmt.Sprintf("https://%s", ctx.Request.Host)))
		log.Println("Notification sent. To get new one reconnect to new stream (restart client app)")
		return true
	})

	ctx.Done()
}

func (m *ctrl) Bind(ctx *gin.Context) {
	m.BindSocket(ctx)
}

func (m *ctrl) BindSocket(ctx *gin.Context) {
	conn, err := wsupgrader.Upgrade(ctx.Writer, ctx.Request, nil)
	if err != nil {
		fmt.Println("Failed to set websocket upgrade: %+v", err)
		return
	}

	log.Println("WebSocket opened. Notification will be sent in 10 seconds")
	time.Sleep(10 * time.Second)
	data,_ := json.Marshal(m.getNotification(fmt.Sprintf("https://%s", ctx.Request.Host)))
	conn.WriteMessage(websocket.TextMessage ,data)
	log.Println("Notification sent. To get new one reconnect to new websocket (restart client app)")
	log.Println("Next notification will be sent in 1 hour if this connection will be opened")
	time.Sleep(1 * time.Hour)
	conn.WriteMessage(websocket.TextMessage ,data)
}

var wsupgrader = websocket.Upgrader{
	ReadBufferSize:  1024,
	WriteBufferSize: 1024,
}

func (m *ctrl) getNotification(currentUrl string) *Update {
	return &Update{
		Icon: fmt.Sprintf("%s%s", currentUrl, "/static/images/unsplash1.jpg"),
		Description: "Based on your experience we found some interesting topics for you",
		Title: "Notifications",
		Link: "https://math-stg.aida.pearson.com/api/v1/assets/videos/467",
	}
}

func NewBinder() endpoints.Binder {
	m := &ctrl{}
	return m
}

endpoints/mission/missionctrl.go
package mission

import (
	"fmt"
	"github.com/gin-gonic/gin"
	"log"
	"net/http"
	"newPocApi/endpoints"
)

type UpdatesMDL struct {
	Updates []Update `json:"updates"`
}

type Update struct {
	Image 	string	`json:"image"`
	Status 	string	`json:"status"`
	Title 	string	`json:"title"`
	Link 	string 	`json:"link"`
}

type ctrl struct {

}

func (m *ctrl) Bind(ctx *gin.Context) {
	ctx.JSON(200, m.getUptates(fmt.Sprintf("https://%s", ctx.Request.Host)))
	if pusher := ctx.Writer.Pusher(); pusher != nil {
		log.Println("Pushing content")
		pusher.Push("/static/images/unsplash1.jpg", &http.PushOptions{
			Method: "GET",
		})
		pusher.Push("/static/images/unsplash2.jpg", &http.PushOptions{
			Method: "GET",
		})
	}

	ctx.Done()
}

func (m *ctrl) getUptates(currentUrl string) *UpdatesMDL {
	return &UpdatesMDL{
		Updates: []Update{
			{
				Image: fmt.Sprintf("%s%s", currentUrl, "/static/images/unsplash1.jpg"),
				Status: "green",
				Title: "Talk about grammar #1",
				Link: "https://math-stg.aida.pearson.com/api/v1/assets/videos/467",
			},{
				Image: fmt.Sprintf("%s%s", currentUrl, "/static/images/unsplash2.jpg"),
				Status: "yellow",
				Title: "Effective presentation #1",
				Link: "https://math-stg.aida.pearson.com/api/v1/assets/videos/467",
			}, {
				Image: fmt.Sprintf("%s%s", currentUrl, "/static/images/unsplash1.jpg"),
				Status: "red",
				Title: "Talk about grammar #2",
				Link: "https://math-stg.aida.pearson.com/api/v1/assets/videos/467",
			},{
				Image: fmt.Sprintf("%s%s", currentUrl, "/static/images/unsplash2.jpg"),
				Status: "yellow",
				Title: "Effective presentation #2",
				Link: "https://math-stg.aida.pearson.com/api/v1/assets/videos/467",
			},
		},
	}
}

func NewBinder() endpoints.Binder {
	m := &ctrl{}
	return m
}